service/history/ndc/mutable_state_mapper.go

```diff
@@ -0,0 +1,182 @@
+		newBufferEventFlusher: newBufferEventFlusher,
+		newBranchMgr:          newBranchMgr,
+		newConflictResolver:   newConflictResolver,
+		newStateBuilder:       newStateBuilder,
+	}
+}
+
+func (m *MutableStateMapperImpl) FlushBufferEvents(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, struct{}, error) {
+	flusher := m.newBufferEventFlusher(wfContext, mutableState, task.getLogger())
+	_, mutableState, err := flusher.flush(ctx)
+	if err != nil {
+		task.getLogger().Error(
+			"MutableStateMapping::FlushBufferEvents unable to flush buffer events",
+			tag.Error(err),
+		)
+		return nil, struct{}{}, err
+	}
+	return mutableState, struct{}{}, err
+}
+
+func (m *MutableStateMapperImpl) PrepareHistoryBranch(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, PrepareHistoryBranchOut, error) {
+	branchMgr := m.newBranchMgr(wfContext, mutableState, task.getLogger())
+	incomingVersionHistory := task.getVersionHistory()
+	doContinue, versionHistoryIndex, err := branchMgr.prepareBranch(
+		ctx,
+		incomingVersionHistory,
+		task.getFirstEvent().GetEventId(),
+		task.getFirstEvent().GetVersion(),
+	)
+	switch err.(type) {
+	case nil:
+		return mutableState, PrepareHistoryBranchOut{
+			DoContinue:  doContinue,
+			BranchIndex: versionHistoryIndex,
+		}, nil
+	case *serviceerrors.RetryReplication:
+		// replication message can arrive out of order
+		// do not log
+		return nil, PrepareHistoryBranchOut{}, err
+	default:
+		task.getLogger().Error(
+			"MutableStateMapping::PrepareHistoryBranch unable to prepare version history",
+			tag.Error(err),
+		)
+		return nil, PrepareHistoryBranchOut{}, err
+	}
+}
+
+func (m *MutableStateMapperImpl) PrepareMutableState(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task PrepareMutableStateIn,
+) (workflow.MutableState, bool, error) {
+	conflictResolver := m.newConflictResolver(wfContext, mutableState, task.getLogger())
+	incomingVersion := task.getVersion()
+	mutableState, isRebuilt, err := conflictResolver.prepareMutableState(
+		ctx,
+		task.BranchIndex,
+		incomingVersion,
+	)
+	if err != nil {
+		task.getLogger().Error(
+			"MutableStateMapping::PrepareMutableState unable to prepare mutable state",
+			tag.Error(err),
+		)
+	}
+	return mutableState, isRebuilt, err
+}
+
+func (m *MutableStateMapperImpl) ApplyEvents(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	task replicationTask,
+) (workflow.MutableState, workflow.MutableState, error) {
+	stateBuilder := m.newStateBuilder(mutableState, task.getLogger())
+	newMutableState, err := stateBuilder.ApplyEvents(
+		ctx,
+		task.getNamespaceID(),
+		uuid.New().String(),
+		*task.getExecution(),
+		task.getEvents(),
+		task.getNewEvents(),
+	)
+	if err != nil {
+		task.getLogger().Error(
```

Comment:
Nit: Seems the caller of this method is also logging the error. Will this produce duplicated error log?

Reply:
good point, let me update